### PR TITLE
[seport] fix the processing of a list of integers as per the ports spec

### PR DESCRIPTION
**Pull request by das7pad**
##### SUMMARY

Port https://github.com/ansible/ansible/pull/58637

> The seport module does not enforce an element type for the ports and the example shows a mixed usage of strings and integers as a list item.
> 
> Using an integer as port breaks the port handling in this module and the handling in the seobject package on the node.
> 
> Casting every port as string fixes the handling.
> 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
seport

##### ADDITIONAL INFORMATION
<details>

```paste below
# traceback in this module
{"changed": false, "module_stderr": "Shared connection to SNIP closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/root/.ansible/tmp/ansible-tmp-1562077098.42-100783222996536/AnsiballZ_seport.py\", line 125, in <module>\r\n    _ansiballz_main()\r\n  File \"/root/.ansible/tmp/ansible-tmp-1562077098.42-100783222996536/AnsiballZ_seport.py\", line 117, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/root/.ansible/tmp/ansible-tmp-1562077098.42-100783222996536/AnsiballZ_seport.py\", line 54, in invoke_module\r\n    imp.load_module('__main__', mod, module, MOD_DESC)\r\n  File \"/tmp/ansible_seport_payload_YvGtLh/__main__.py\", line 307, in <module>\r\n  File \"/tmp/ansible_seport_payload_YvGtLh/__main__.py\", line 297, in main\r\n  File \"/tmp/ansible_seport_payload_YvGtLh/__main__.py\", line 206, in semanage_port_add\r\n  File \"/tmp/ansible_seport_payload_YvGtLh/__main__.py\", line 159, in semanage_port_get_type\r\nAttributeError: 'int' object has no attribute 'split'\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}

# traceback in seobject, string handling fixed in the semanage_port_get_type function
{"changed": false, "module_stderr": "Shared connection to SNIP closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/root/.ansible/tmp/ansible-tmp-1562077932.73-31177858006091/AnsiballZ_seport.py\", line 125, in <module>\r\n    _ansiballz_main()\r\n  File \"/root/.ansible/tmp/ansible-tmp-1562077932.73-31177858006091/AnsiballZ_seport.py\", line 117, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/root/.ansible/tmp/ansible-tmp-1562077932.73-31177858006091/AnsiballZ_seport.py\", line 54, in invoke_module\r\n    imp.load_module('__main__', mod, module, MOD_DESC)\r\n  File \"/tmp/ansible_seport_payload_Iz0NRY/__main__.py\", line 310, in <module>\r\n  File \"/tmp/ansible_seport_payload_Iz0NRY/__main__.py\", line 300, in main\r\n  File \"/tmp/ansible_seport_payload_Iz0NRY/__main__.py\", line 213, in semanage_port_add\r\n  File \"/usr/lib64/python2.7/site-packages/seobject/__init__.py\", line 1200, in modify\r\n    self.__modify(port, proto, serange, setype)\r\n  File \"/usr/lib64/python2.7/site-packages/seobject/__init__.py\", line 1168, in __modify\r\n    (k, proto_d, low, high) = self.__genkey(port, proto)\r\n  File \"/usr/lib64/python2.7/site-packages/seobject/__init__.py\", line 1074, in __genkey\r\n    ports = port.split(\"-\")\r\nAttributeError: 'int' object has no attribute 'split'\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```
</details>

### CODE REVIEW

**Comment by felixfontein** [comment]
Why not simply add `elements='str'` to `ports=dict(type='list', required=True),` and let Ansible make sure that the list elements are strings?

(Also, this needs a changelog fragment.)

**Comment by das7pad** [comment]
> Why not simply add `elements='str'` to `ports=dict(type='list', required=True),` and let Ansible make sure that the list elements are strings?

The docs [1] are not very specific on whether there will be type casting or whether this will error out when specifying a different element type -- e.g. a list of int like in the examples.

Are there unit tests that show the behavior here?

----
 [1] https://docs.ansible.com/ansible/latest/dev_guide/developing_program_flow_modules.html#elements

**Comment by felixfontein** [comment]
`elements` behaves the same as `type`, i.e. it will typecast if possible, and fail it it's not possible. There are probably unit tests somewhere, check out https://github.com/ansible/ansible/tree/devel/test/units/module_utils/basic and/or https://github.com/ansible/ansible/tree/devel/test/units/module_utils/common.

**Comment by Andersson007** [comment]
any updates on this?
needs_info

File: plugins/modules/system/seport.py [hunk 1]
```diff
@@ -278,7 +278,7 @@ def main():
     if not get_runtime_status(ignore_selinux_state):
         module.fail_json(msg="SELinux is disabled on this host.")
 
-    ports = module.params['ports']
+    ports = list(map(str, module.params['ports']))
```
Comment (russoz) [inline plugins/modules/system/seport.py:281]: It might be slightly simpler as:
```
ports = [str(p) for p in ports]
```

Comment (felixfontein) [reply plugins/modules/system/seport.py:281]: `elements='str'` in the argument spec is even more simple :)

**Comment by das7pad** [comment]
> `elements` behaves the same as `type`, i.e. it will typecast if possible, and fail it it's not possible. There are probably unit tests somewhere, check out https://github.com/ansible/ansible/tree/devel/test/units/module_utils/basic and/or https://github.com/ansible/ansible/tree/devel/test/units/module_utils/common.

Thanks, I found one:

https://github.com/ansible/ansible/blob/1a2da990a403e1c836574c5883ec0c17a98d2a65/test/units/module_utils/basic/test_argument_spec.py#L65-L66
```py
    # Type=list elements=str implicit converts to string
    ({'arg': {'type': 'list', 'elements': 'str'}}, {'arg': [42, 32]}, ['42', '32']),
```

**Comment by felixfontein** [comment]
I just wanted to review this, when I noticed that this has already been changed in `main`: https://github.com/ansible-collections/community.general/blob/main/plugins/modules/system/seport.py#L262 Probably in one of the bulk ignore.txt cleanups.

Sorry, I guess this PR won't be needed anymore :/

(I think GitHub doesn't show this as a conflict since the change is now exactly as in `main`, so rebasing/merging is possible without human intervention...)

**Comment by das7pad** [comment]
> Sorry, I guess this PR won't be needed anymore :/

No worries. I'm closing this then.

**Comment by felixfontein** [comment]
@das7pad thanks for working on this!

**Pull request closed without merging**